### PR TITLE
Add ICU 75.1

### PR DESCRIPTION
**Pull request by datalogics-staylor**
ICU version 75.1 was just released.

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Review by jwillikers — COMMENTED** [review]
I'm curious whether we need the mingw patch or not for this release.

File: recipes/icu/all/conandata.yml [hunk 1]
```diff
@@ -27,6 +30,8 @@ sources:
     url: "https://github.com/unicode-org/icu/releases/download/release-68-2/icu4c-68_2-src.tgz"
     sha256: "c79193dee3907a2199b8296a93b52c5cb74332c26f3d167269487680d479d625"
 patches:
+  "75go.1":
```
Comment (jwillikers) [inline recipes/icu/all/conandata.yml:33]:
```suggestion
  "75.1":
```

**Review by jwillikers — CHANGES_REQUESTED** [review]

File: recipes/icu/all/conandata.yml (hunk 1 above)
Comment (jwillikers) [reply recipes/icu/all/conandata.yml]: It happens. I usually blame my code editor, whether or not that's what actually caused it ;)

**Review by jwillikers — APPROVED** [review]

**Comment by jwillikers** [comment]
According to the [release announcement](https://icu.unicode.org/):

> C++ code now requires C++17 and is being made more robust. 

Looks like that constraint will need to be enforced for version 75 and later. You should be able to pull the logic from the Qt 6 recipe to require this in the `validate` method.

**Comment by datalogics-staylor** [comment]
> According to the [release announcement](https://icu.unicode.org/):
> 
> > C++ code now requires C++17 and is being made more robust.
> 
> Looks like that constraint will need to be enforced for version 75 and later. You should be able to pull the logic from the Qt 6 recipe to require this in the `validate` method.

Makes sense. Thanks and I will do that!

File: recipes/icu/all/conanfile.py [hunk 2]
```diff
@@ -75,6 +75,9 @@ def validate(self):
             if not os.path.exists(str(self.options.dat_package_file)):
                 raise ConanInvalidConfiguration("Non-existent dat_package_file specified")
 
+        if self.settings.compiler.get_safe("cppstd"):
```
Comment (jwillikers) [inline recipes/icu/all/conanfile.py:78]: You'll probably also need something like [this](https://github.com/conan-io/conan-center-index/blob/7c80e9e24abaa4963b6cd42eb7dab023d4d256a2/recipes/qt/6.x.x/conanfile.py#L188-L197) and [this](https://github.com/conan-io/conan-center-index/blob/7c80e9e24abaa4963b6cd42eb7dab023d4d256a2/recipes/qt/6.x.x/conanfile.py#L258-L262). That enforces an additional check that the minimum version of the compiler supports C++17.

Comment (datalogics-staylor) [reply recipes/icu/all/conanfile.py:78]: Thank you for all the help! In case it wasn't obvious, this aspect of conan packages is fairly new to me.

Comment (jwillikers) [reply recipes/icu/all/conanfile.py:78]: No problem. It's always good to see a new contributor helping out, and I'm looking forward to the new version. I should also point you to the CMake package template, [here](https://github.com/conan-io/conan-center-index/blob/master/docs/package_templates/cmake_package/all/conanfile.py), which you'll probably find really useful.

**Comment by jwillikers** [comment]
Looks like it is just breaking on the `test_package` now.

**Comment by datalogics-staylor** [comment]
> Looks like it is just breaking on the `test_package` now, but only for Conan V1.

Yes, I've been trying to figure out what to do about those failures and I'm afraid I'm coming up short.

File: recipes/icu/all/conanfile.py [hunk 3]
```diff
@@ -58,6 +58,16 @@ def _enable_icu_tools(self):
...
         return not self.conf.get("tools.build:skip_test", default=True, check_type=bool)
 
+    @property
+    def _minimum_compilers_version(self):
+        return {
+            "Visual Studio": "16",
+            "msvc": "192",
+            "gcc": "8",
+            "clang": "9",
+            "apple-clang": "12" if Version(self.version) >= "6.5.0" else "11"
```
Comment (jwillikers) [inline recipes/icu/all/conanfile.py:68]: This just needs adapted from the Qt 6 recipe.

```suggestion
            "apple-clang": "11"
```

**Review by jwillikers — CHANGES_REQUESTED** [review]
@datalogics-staylor The issue is due to the `test_package` not having a C++ standard set when being built with Conan V1. You need to add `target_compile_features(${PROJECT_NAME} PRIVATE cxx_std_17)` to the `CMakeLists.txt` in the `test_package` if the version of ICU is greater 75 or later.

Something like the following will probably work.

```
if (ICU_VERSION VERSION GREATER_EQUAL "75")
    target_compile_features(${PROJECT_NAME} PRIVATE cxx_std_17)
else()
    target_compile_features(${PROJECT_NAME} PRIVATE cxx_std_11)
endif()
```

File: recipes/icu/all/conanfile.py (hunk 3 above)
Comment (datalogics-staylor) [reply recipes/icu/all/conanfile.py:68]: Thank you so much! This makes perfect sense now that I understand what went wrong. I was having trouble understanding what the difference from Conan 1 to Conan 2 causing this was from just the logs.

**Comment by jwillikers** [comment]
MSVC isn't using the C++17 standard:

> note: 'auto' in non-type template parameters requires at least '/std:c++17'

@RubenRBS What's the best way to go about this?

**Pull request closed without merging**